reflection/serverreflection.go

```diff
@@ -37,319 +37,194 @@ To register server reflection on a gRPC server:
+	return svcInfoFromNames(serviceNames)
 }
 
-func fqn(prefix, name string) string {
-	if prefix == "" {
-		return name
-	}
-	return prefix + "." + name
-}
-
-// decodeFileDesc does decompression and unmarshalling on the given
-// file descriptor byte slice.
-func decodeFileDesc(enc []byte) (*dpb.FileDescriptorProto, error) {
-	raw, err := decompress(enc)
-	if err != nil {
-		return nil, fmt.Errorf("failed to decompress enc: %v", err)
-	}
+type svcInfoFromNames []string
 
-	fd := new(dpb.FileDescriptorProto)
-	if err := proto.Unmarshal(raw, fd); err != nil {
-		return nil, fmt.Errorf("bad descriptor: %v", err)
+func (s svcInfoFromNames) GetServiceInfo() map[string]grpc.ServiceInfo {
+	m := make(map[string]grpc.ServiceInfo, len(s))
+	for _, name := range s {
+		m[name] = grpc.ServiceInfo{}
 	}
-	return fd, nil
+	return m
 }
 
-// decompress does gzip decompression.
-func decompress(b []byte) ([]byte, error) {
-	r, err := gzip.NewReader(bytes.NewReader(b))
-	if err != nil {
-		return nil, fmt.Errorf("bad gzipped descriptor: %v", err)
-	}
-	out, err := ioutil.ReadAll(r)
-	if err != nil {
-		return nil, fmt.Errorf("bad gzipped descriptor: %v", err)
-	}
-	return out, nil
+// Register registers the server reflection service on the given gRPC server.
+func Register(s GRPCServer) {
+	svr := NewServer(ServerOptions{Services: s})
+	rpb.RegisterServerReflectionServer(s, svr)
 }
 
-func fileDescContainingExtension(typeName string, ext int32) (*dpb.FileDescriptorProto, error) {
-	desc, err := protoregistry.GlobalFiles.FindDescriptorByName(protoreflect.FullName(typeName))
-	if err != nil {
-		return nil, err
-	}
-	m := dynamicpb.NewMessage(desc.(protoreflect.MessageDescriptor))
-
-	var extDesc *proto.ExtensionDesc
-	for id, desc := range proto.RegisteredExtensions(m) {
-		if id == ext {
-			extDesc = desc
-			break
+func (s *serverReflectionServer) init() {
+	s.initServiceNames.Do(func() {
+		if s.s == nil {
+			// no need to init; no service names advertised
+			return
 		}
-	}
-
-	if extDesc == nil {
-		return nil, fmt.Errorf("failed to find registered extension for extension number %v", ext)
-	}
-
-	return decodeFileDesc(proto.FileDescriptor(extDesc.Filename))
+		serviceInfo := s.s.GetServiceInfo()
+		s.serviceNames = make([]string, 0, len(serviceInfo))
+		for svc := range serviceInfo {
+			s.serviceNames = append(s.serviceNames, svc)
+		}
+		sort.Strings(s.serviceNames)
+	})
 }
 
 // fileDescWithDependencies returns a slice of serialized fileDescriptors in
 // wire format ([]byte). The fileDescriptors will include fd and all the
 // transitive dependencies of fd with names not in sentFileDescriptors.
-func fileDescWithDependencies(fd *dpb.FileDescriptorProto, sentFileDescriptors map[string]bool) ([][]byte, error) {
-	r := [][]byte{}
-	queue := []*dpb.FileDescriptorProto{fd}
+func (s *serverReflectionServer) fileDescWithDependencies(fd protoreflect.FileDescriptor, sentFileDescriptors map[string]struct{}) ([][]byte, error) {
```

Comment:
On `map[string]bool` vs `map[string]struct{}`.
1. I'm not sure which is more memory efficient, but I think the difference is small enough that we wouldn't care
2. In gRPC-Go, we've mostly (but not as an official policy) adopted `map[string]bool` as "set"s. Since if the entry doesn't exist, the map read returns the default bool (which is false). And we don't need to write `struct{}{}` when adding entries.

Let me know what you think.

Reply:
Done